plugin-base/src/main/java/io/everytrade/server/plugin/impl/everytrade/parser/exchange/BinanceExchangeSpecificParserV4.java

```diff
@@ -0,0 +1,116 @@
+    public BinanceExchangeSpecificParserV4(Class<? extends ExchangeBean> exchangeBean) {
+        super(exchangeBean);
+    }
+
+    List<BinanceBeanV4> rows;
+    List<BinanceBeanV4> unSupportedRows = new ArrayList<>();
+
+    public List<? extends ExchangeBean> convertMultipleRowsToTransactions(List<BinanceBeanV4> rows) {
+        List<BinanceBeanV4> result;
+        this.rows = setRowsWithIds(rows);
+        var groupedRowsByTime = this.rows.stream().collect(groupingBy(BinanceBeanV4::getDate));
+        Map<Instant, List<BinanceBeanV4>> sortedGroupsByDate = new TreeMap<>(groupedRowsByTime);
+        // merging rows nearly in the same time
+        var mergedGroups = mergeGroupsInTimeWithinTolerance(sortedGroupsByDate);
+        // clean groups of rows from unsupported rubbish
+        var cleanGroups = purgeGroupsWithUnsupportedRows(mergedGroups);
+        // creating transaction
+        List<BinanceBeanV4> rowsReadyForTxs = createTransactionFromGroupOfRows(cleanGroups);
+        result = rowsReadyForTxs;
+        result.addAll(unSupportedRows);
+        return rowsReadyForTxs;
+    }
+
+    private Map<Instant, List<BinanceBeanV4>> purgeGroupsWithUnsupportedRows(Map<Instant, List<BinanceBeanV4>> rowGroups) {
+        Map<Instant, List<BinanceBeanV4>> result = new HashMap<>();
+        for (Map.Entry<Instant, List<BinanceBeanV4>> entry : rowGroups.entrySet()) {
+            var rowsInGroup = entry.getValue();
+            var isOneOrMoreUnsupportedRows =
+                !rowsInGroup.stream().filter(r -> r.isUnsupportedRow() == true).collect(Collectors.toList()).isEmpty();
+            if (!isOneOrMoreUnsupportedRows) {
+                result.put(entry.getKey(), entry.getValue());
+            } else {
+                setRowsAsUnsupported(rowsInGroup, "One or more rows in group are unsupported");
+            }
+        }
+        return result;
+    }
+
+    private void setRowsAsUnsupported(List<BinanceBeanV4> rowsInGroup, String message) {
+        rowsInGroup.forEach(r -> {
+            r.setMessage(message);
+            r.setUnsupportedRow(true);
+        });
+        unSupportedRows.addAll(rowsInGroup);
+    }
+
+    private List<BinanceBeanV4> setRowsWithIds(List<BinanceBeanV4> rows) {
+        int i = 1;
+        for (BinanceBeanV4 row : rows) {
+            i++;
+            row.setRowId(i);
+        }
+        return rows;
+    }
+
+    private Map<Instant, List<BinanceBeanV4>> mergeGroupsInTimeWithinTolerance(Map<Instant, List<BinanceBeanV4>> groups) {
+        Map<Instant, List<BinanceBeanV4>> result = new HashMap<>();
+        Instant previousKey = Instant.EPOCH;
+        List<BinanceBeanV4> previousValues = new ArrayList<>();
+        for (Map.Entry<Instant, List<BinanceBeanV4>> entry : groups.entrySet()) {
+            var currentKey = entry.getKey();
+            var currentValues = entry.getValue();
+            if ((currentKey.minusMillis(TRANSACTION_MERGE_TOLERANCE).equals(previousKey)
```

Comment:
is this correct? I would expect both side tolerance something like :
`currentKey.minusMillis(TRANSACTION_MERGE_TOLERANCE).isBefore(previousKey) && currentKey.plusMillis(TRANSACTION_MERGE_TOLERANCE).isAfter(previousKey)`

so that previous key is in interval currentKey plus minus TOLERANCE

Reply:
Yes, it is.